mitiq/zne/scaling/identity_insertion.py

```diff
@@ -0,0 +1,425 @@
+        new_output_circuit_weight = output_circuit_weight + weight_mask[j]
+        new_approx_error = np.abs(
+            new_output_circuit_weight - scale_factor * input_circuit_weight
+        )
+        # Fold the candidate gate only if it helps improving the approximation
+        if new_approx_error >= approx_error:
+            break
+        approx_error = new_approx_error
+        output_circuit_weight = new_output_circuit_weight
+        num_inserts_mask[j] += 1
+
+    return num_inserts_mask
+
+
+def _apply_scale_mask(
+    circuit: Circuit,
+    num_inserts_mask: List[int],
+    squash_moments: Optional[bool] = True,
+) -> Circuit:
+    r"""Applies local identity scaling to the gates of the input circuit
+    according to the input num_folds_mask.
+    More precicely, I is applied after the j_th gate G_j of
+    the input circuit an integer number of times given by num_folds_mask[j].
+    The gate ordering is equal to the one used in the `all_operations()`
+    method of the :class:`cirq.Circuit` class: gates from earlier moments
+    come first and gates within the same moment follow the order in which
+    they were given to the moment's constructor.
+    Args:
+        circuit: The quantum circuit to scale.
+        num_inserts_mask: The list of integers indicating how many times
+            the corresponding gates of 'circuit' should be folded.
+        squash_moments: If True or None, all gates (including inserted gates)
+            are placed as early as possible in the circuit. If False, new
+            moments are created for inserted gates. This option only applies to
+            QPROGRAM types which have a "moment" or "time" structure. Default
+            is True.
+    Returns: The scaled quantum circuit.
+    """
+    _check_scalable(circuit)
+    circ_copy = deepcopy(circuit)
+    measurements = _pop_measurements(circ_copy)
+
+    num_gates = len(list(circ_copy.all_operations()))
+    if num_gates != len(num_inserts_mask):
+        raise ValueError(
+            "The circuit and the scaling mask have incompatible sizes."
+            f" The number of gates is {num_gates}"
+            f" but len(num_folds_mask) is {len(num_inserts_mask)}."
+        )
+
+    scaled_circuit = circ_copy[:0]
+    if squash_moments:
+        for op, num_inserts in zip(
+            circ_copy.all_operations(), num_inserts_mask
+        ):
+            scaled_circuit.append(
+                [op]
+                + num_inserts
+                * [GateOperation(IdentityGate(len(op.qubits)), op.qubits)]
+            )
+    else:
+        mask_index = 0
+        for moment in circ_copy:
+            scaled_moment = Circuit(moment)
+            for op in moment:
+                num_inserts = num_inserts_mask[mask_index]
+                scaled_moment.append(
+                    num_inserts
+                    * [GateOperation(IdentityGate(len(op.qubits)), op.qubits)]
+                )
```

Comment:
This function is identical to [`_apply_fold_mask`](https://github.com/unitaryfund/mitiq/blob/master/mitiq/zne/scaling/folding.py#L491) except for these lines which determine which gates are inserted. Here, it is identity, and in unitary folding it is an operation in the circuit.

I would propose generalizing `_apply_fold_mask` to the following:

```python
def _apply_fold_mask(
    circuit: Circuit,
    num_folds_mask: List[int],
    squash_moments: Optional[bool] = True,
    gates_to_insert: Optional[Sequence[cirq.Gate]] = None,
) -> Circuit:
    """...

    Args:
        ...
        gates_to_insert: Pool of gates to select from and insert into the circuit when applying the fold mask.
            When inserting a gate, we choose `G` uniformly at random from `gates_to_insert` and insert
            G^dag G into the circuit. If `None`, `G` is the preceding operation in the circuit.
    """
```

and using it here with `gates_to_insert = cirq.I`.

In addition to reducing duplication, this would increase the flexibility of folding/scaling and would allow us to test more methods. Thoughts @andreamari? 

Some considerations:

- Gates might need to be specified in a frontend-agnostic manner (e.g., `"Identity"` or `"I"` instead of `cirq.I`.)
- Might want to allow for a probability distribution (not just uniform as I wrote above) for selecting from `gates_to_insert`.
- For the inserted gate, we probably want to match the number of qubits of the preceding gate in the circuit.

Reply:
In unitary folding, there's 2 operations being inserted which is why I created this separate function. I could try to figure out if there's a way to change just those two lines when `apply_fold_mask` is used for identity scaling. I think there might be a way to define a decorator function. 

> For the inserted gate, we probably want to match the number of qubits of the preceding gate in the circuit.

I think it does so from what I remember. Some of the tests are for gates acting on more than 1 qubit and the gates inserted also act on the same number of qubits. I will change the docstrings wherever possible. 

> Gates might need to be specified in a frontend-agnostic manner 

Ok. Seems reasonable. I will make the changes. 

> Might want to allow for a probability distribution (not just uniform as I wrote above) for selecting from gates_to_insert

I think this could be added to above decorator function.